### ISSUE TITLE
Compile HLG runs OOM and fails

### ISSUE THREAD

**Issue by gindigo** [k2-fsa/k2#1029]
Hi,
I'm adapting Librispeech scripts to other datasets I have access to. I'm facing a problem on stage 9 "Compile HLG". After calling "./local/compile_hlg.py --lang-dir data/lang_phone" it stays in the step "Determinizing LG" until the machine runs OOM and the process is killed.
This is what I see:
`2022-08-01 14:51:02 (prepare.sh:215:main) Stage 9: Compile HLG
2022-08-01 14:51:03,336 INFO [compile_hlg.py:148] Processing data/lang_phone
2022-08-01 14:51:03,700 INFO [lexicon.py:179] Converting L.pt to Linv.pt
2022-08-01 14:51:04,363 INFO [compile_hlg.py:64] Building ctc_topo. max_token_id: 40
2022-08-01 14:51:04,504 INFO [compile_hlg.py:76] Loading G_3_gram.fst.txt
2022-08-01 14:51:06,178 INFO [compile_hlg.py:87] Intersecting L and G
2022-08-01 14:51:10,908 INFO [compile_hlg.py:89] LG shape: (3892518, None)
2022-08-01 14:51:10,908 INFO [compile_hlg.py:91] Connecting LG
2022-08-01 14:51:10,908 INFO [compile_hlg.py:93] LG shape after k2.connect: (3892518, None)
2022-08-01 14:51:10,908 INFO [compile_hlg.py:95] <class 'torch.Tensor'>
2022-08-01 14:51:10,908 INFO [compile_hlg.py:96] Determinizing LG
`
I'm using a lexicon and 3gr/4gr LMs which work in kaldi (older version). I have verified that the words in the LMs exist in the lexicon, so no OOVs. The LMs were trained using srilm toolbox(ngram-count) -with some smoothing/discount flags. Unfortunately I don't have the original training material used to train such LMs.

I suspect the reported issue concerns some characteristics of the LMs, but I'm not sure where to look at. 

Then I tried the same with a different data. Again if I used my old LMs, the HLG compilation fails. However, if I retrain the LMs without smoothing/discount flags,  then the HLG is compiled without problems.

I was hoping to adjust the k2 scripts and reuse my existing LMs, so I'd appreciate your input on this.  Are there any LM requirements/validations I could check for my existing LMs? Have you encountered a similar issue and been able to adjust the LMs in some way?

Thanks for your feedback.

**Comment by danpovey** [comment]
Our determinization algorithm may not be as memory-efficient as the one in openfst.  I assume your LMs are on the larger side or you don't have tons of CPU memory?

@csukuangfj I wonder whether, for purposes of creating offline training graphs, we should try to create a version of this script that uses your wrapping of OpenFST instead?

**Comment by csukuangfj** [comment]
> @csukuangfj I wonder whether, for purposes of creating offline training graphs, we should try to create a version of this script that uses your wrapping of OpenFST instead?

Yes, I agree. Several people have encountered such an OOM issue due to using large LMs.
Will try to implement an equivalent Python version using OpenFST.

**Comment by gindigo** [comment]
Thanks for your insights. 
Yes, these are not very-small LMs. In case it is of any help, I'm reporting here more details.

In my previous post, I mentioned two cases. These are the elements of the case in which HLG compiling fails:
Lex: 277K words
3gram (14MB):
ngram 1=257745
ngram 2=170194
ngram 3=164955

4gram (15MB):
ngram 1=257745
ngram 2=168735
ngram 3=120863
ngram 4=68012

And these are the elements of the case that works fine (after retraining LMs withot discounts/pruning)
Lex: 175K words
3gram (66MB):
ngram 1=161130
ngram 2=1689052
ngram 3=908925

4gram (88MB):
ngram 1=161130
ngram 2=1689052
ngram 3=908925
ngram 4=648269

Machine:
(AWS g3.8xlarge) 32xCPU 244GxMem 2xGPU-(2x8G gpuMem)

**Comment by danpovey** [comment]
OK.  After a certain point, the additional RAMy doesn't help me because k2 uses 32-bit integer indexes.  This saves, on average, nearly a factor of 2 in memory, but it limits now large FSAs can be.  I believe OpenFST does the same, although it may be configurable.  Anyway they use different algorithms so just because k2 fails, doesn't necessarily mean that OpenFST would fail.
